modules/hdf5_handler/HDF5VlenAtomicArray.cc

```diff
@@ -0,0 +1,262 @@
+    if (memspace < 0) {
+        H5Dclose(dset_id);
+        H5Fclose(file_id);
+        throw InternalErr(__FILE__, __LINE__, "could not create data space");
+    }
+
+    vector<hvl_t> vlen_data(vlen_num_elems);
+    if (H5Dread(dset_id, vlen_memtype, memspace, vlen_space, H5P_DEFAULT, vlen_data.data()) <0) {
+        H5Dclose(dset_id);
+        H5Fclose(file_id);
+        throw InternalErr(__FILE__, __LINE__,"Cannot read variable-length datatype data.");
+    }
+
+    // Handle vlen and vlen_index here.
+    if (vlen_index) {
+        if (this->var()->type() != dods_int32_c) { 
+            H5Dclose(dset_id); 
+            H5Fclose(file_id);
+            string err_msg = "vlen_index datatype must be 32-bit integer."; 
+            throw InternalErr( __FILE__, __LINE__,err_msg); 
+        } 
+        vector<int> vlen_index_data; 
+        for (ssize_t i = 0; i<vlen_num_elems; i++)  
+            vlen_index_data.push_back(vlen_data[i].len); 
+        set_value_ll(vlen_index_data.data(),vlen_num_elems); 
+    }
+    else {
+
+        switch (this->var()->type()) {
+            case dods_byte_c:
+            case dods_uint8_c:
+            case dods_char_c:
+            case dods_int8_c:
+            case dods_int16_c:
+            case dods_uint16_c:
+            case dods_int32_c:
+            case dods_uint32_c:
+            case dods_int64_c:
+            case dods_uint64_c:
+            case dods_float32_c:
+            case dods_float64_c: {
+
+                // Retrieve the last dimension size.
+                libdap::Array::Dim_iter last_dim_iter = this->dim_end()-1;
+                int64_t last_dim_size = this->dimension_size(last_dim_iter);
+                size_t bytes_per_element = this->var()->width_ll();
+                size_t total_data_buf_size = vlen_num_elems*last_dim_size*bytes_per_element;
+
+                vector<char> data_buf(total_data_buf_size,0);
+                char *temp_data_buf_ptr = data_buf.data();
+
+                for (ssize_t i = 0; i < vlen_num_elems; i++) {
+
+                    size_t vlen_element_size = vlen_data[i].len * bytes_per_element;
+
+                    // Copy the vlen data to the data buffer.
+                    memcpy(temp_data_buf_ptr,vlen_data[i].p,vlen_element_size);
+
+                    // Move the data buffer pointer to the next element.
+                    // In this regular array, the rest data will be filled with zero.
+                    temp_data_buf_ptr += last_dim_size*bytes_per_element;
+
+                }
+                val2buf(data_buf.data());
+
+                break;
+            }
+            default:
+                throw InternalErr(__FILE__, __LINE__, "Vector::val2buf: bad type");
+        }
+    }
+    if(H5Dvlen_reclaim(vlen_memtype, memspace, H5P_DEFAULT, (void*)(vlen_data.data())) <0) {
+        H5Dclose(dset_id);
+        H5Fclose(file_id);
+        throw InternalErr(__FILE__, __LINE__, "H5Dvlen_reclaim failed.");
+    }
+    H5Sclose(vlen_space);
```

Comment:
It's hard to tell from the diff, but are all these taken care of when exceptions are thrown?

Reply:
Actually no. In theory, they should. I agree they should be handled. Will do.

Reply:
Just checked in the update.